src/plugins/discord/fetcher.js

```diff
@@ -0,0 +1,149 @@
+// @flow
+
+import * as Model from "./models";
+import * as NullUtil from "../../util/null";
+
+type FetchEndpoint = (endpoint: string) => Promise<any>;
+
+/**
+ * Provide the Guild ID to fetch against, and the 'limit'
+ * parameter when fetching GuildMembers, Messages, and Reactions.
+ *
+ */
+type FetchOptions = {|
+  guildId: Model.Snowflake,
+  membersLimit: number,
+  messagesLimit: number,
+  reactionsLimit: number,
+|};
+
+export type EndCursor = Model.Snowflake | null;
+
+export type PageInfo = {|
+  +hasNextPage: boolean,
+  +endCursor: EndCursor,
+|};
+
+export type ResultPage<T> = {|
+  +pageInfo: PageInfo,
+  +results: $ReadOnlyArray<T>,
+|};
+
+/**
+ * Fetcher is responsible for:
+ * - Returning the correct endpoint to fetch against for Guilds, Channels,
+ *   Members, and Reactions.
+ * - Formatting the returned results into the correct Typed objects
+ * - Returning pagination info in a PageInfo object, containing hasNextPage
+ *   and endCursor properties.
+ *   The endCursor property is calculated as the Id of the last object
+ *   returned in the response results. We are assuming Discord provides
+ *   consistent, ordered results.
+ *   The hasNextPage property is a boolean calculated as whether the number of
+ *   results recieved is equal to the `limit` property provided in the
+ *   fetch request.
+ *
+ *   Note that Discord doesn't support pagination for Channels, so we're
+ *   returning an array of Channel objects in the corresponding method.
+ *   See: https://discordapp.com/developers/docs/resources/guild#get-guild-channels
+ *
+ */
+export class Fetcher {
+  +_fetch: FetchEndpoint;
+  +_options: FetchOptions;
+
+  constructor(fetchEndpoint: FetchEndpoint, options: FetchOptions) {
+    this._fetch = fetchEndpoint;
+    this._options = options;
+  }
+
+  async guild(): Promise<Model.Guild> {
+    const {guildId} = this._options;
+    const {id, name, permissions} = await this._fetch(`guilds/${guildId}`);
+    return {id: id, name: name, permissions: permissions};
+  }
+
+  async channels(): Promise<$ReadOnlyArray<Model.Channel>> {
+    const {guildId} = this._options;
+    const response = await this._fetch(`/guilds/${guildId}/channels`);
+    return response.map((x) => ({
+      id: x.id,
+      name: x.name,
+      type: Model.channelTypeFromId(x.type),
+    }));
+  }
+
+  async members(after: EndCursor): Promise<ResultPage<Model.GuildMember>> {
+    const {membersLimit, guildId} = this._options;
+    const endpoint = `/guilds/${guildId}/members?after=${NullUtil.get(
+      after
+    )}&limit=${membersLimit}`;
```

Comment:
Rather than erroring when `after==null`, we could just only append the `after` query slug when the cursor is not null, right?

Reply:
I think we'd always want an after parameter; but this was obsoleted by updating the `EndCursor`. These endpoints are easier on the eyes now.